File: src/cua/handoff/recorder.py
"""Recording what the human actually did to the live session.

Two sources, because neither alone is complete:

* in-page capture-phase listeners, which see clicks, field changes and form
  submits with the control's accessible name;
* frame-URL diffing from the Python side, which catches navigations that the
  in-page listeners cannot report because the document they lived in was
  replaced mid-event.

Every recorded action is passed through the same redactor as the rest of the
run before it is persisted.
"""

from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path

from ..safety.redact import Redactor
from ..surface.web import PlaywrightWebSurface
from .control import HumanAction

_RECORDER_JS = (Path(__file__).parent / "recorder.js").read_text()
_DRAIN_JS = """() => {
  let t = window; try { t = window.top || window; } catch (e) { t = window; }
  const out = t.__cua_human || [];
  t.__cua_human = [];
  return out;
}"""


def _ts() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


class HumanActionRecorder:
    def __init__(self, surface: PlaywrightWebSurface, redactor: Redactor, *, actor: str = "human",
                 simulated: bool = False) -> None:
        self.surface = surface
        self.redactor = redactor
        self.actor = actor
        self.simulated = simulated
        self._frame_urls: dict[str, str] = {}

    def install(self) -> None:
        """Idempotent; re-run after every navigation to cover new documents."""
        try:
            self.surface.page.evaluate(_RECORDER_JS)
        except Exception:
            pass
        self._frame_urls = self._current_frame_urls()

    def _current_frame_urls(self) -> dict[str, str]:
        urls: dict[str, str] = {}
        try:
            for frame in self.surface.page.frames:
                urls[frame.name or frame.url] = frame.url
        except Exception:
            pass
        return urls

    def drain(self) -> list[HumanAction]:
        actions: list[HumanAction] = []

        # Navigations first: they explain why the in-page buffer may be short.
        current = self._current_frame_urls()
        for key, url in current.items():
            before = self._frame_urls.get(key)
            if before is not None and before != url:
                actions.append(
                    HumanAction(
                        at=_ts(),
                        kind="navigation",
                        description=f"frame {key!r} moved to a new page",
                        frame_url=self.redactor.scrub_text(url),
                        actor=self.actor,
                        simulated=self.simulated,
                    )
                )
        self._frame_urls = current

        raw = []
        try:
            raw = self.surface.page.evaluate(_DRAIN_JS) or []
        except Exception:
            raw = []

        for rec in raw:
            kind = rec.get("kind", "event")
            name = rec.get("name") or rec.get("text") or "(unnamed control)"
            if kind == "click":
                desc = f"clicked the {name!r} {rec.get('role', 'control')}"
            elif kind == "input":
                desc = f"entered a value into {name!r}"
            elif kind == "submit":
                desc = f"submitted form {name!r}"
            else:
                desc = f"{kind} on {name!r}"
            coords = (
                [float(rec["x"]), float(rec["y"])]
                if rec.get("x") is not None and rec.get("y") is not None
                else None
            )
            actions.append(
                HumanAction(
                    at=rec.get("at") or _ts(),
                    kind=kind,
                    description=self.redactor.scrub_text(desc),
                    frame_url=self.redactor.scrub_text(rec.get("frame_url", "")),
                    value=(
                        self.redactor.scrub_text(str(rec["value"]))
                        if rec.get("value") is not None
                        else None
                    ),
                    coordinates=coords,
                    actor=self.actor,
                    simulated=self.simulated,
                )
            )

        # Re-attach for any document created since the last drain.
        self.install()
        return actions

File: src/cua/handoff/recorder.js
/*
 * Human action recorder.
 *
 * Installed into the SAME live page the automation drives, across every
 * same-origin frame. Capture-phase listeners mean the recorder sees the event
 * before the application's own handlers can stop propagation, so a legacy
 * onclick that navigates away is still recorded.
 *
 * Password field values are never captured, at the source.
 */
(function () {
  var top = window;
  try { top = window.top || window; } catch (e) { top = window; }
  if (!top.__cua_human) top.__cua_human = [];

  function txt(el) {
    if (!el) return "";
    var t = (el.innerText || el.textContent || "").replace(/\s+/g, " ").trim();
    return t.length > 120 ? t.slice(0, 120) : t;
  }
  function at(el, n) { try { return el.getAttribute(n) || ""; } catch (e) { return ""; } }

  function describe(el) {
    if (!el || !el.tagName) return { role: "unknown", name: "", secret: false };
    var tag = el.tagName.toLowerCase();
    var type = (at(el, "type") || "").toLowerCase();
    var role =
      tag === "a" ? "link"
      : tag === "select" ? "combobox"
      : tag === "button" ? "button"
      : tag === "input"
        ? (type === "submit" || type === "button" || type === "reset" ? "button" : "textbox")
        : (at(el, "onclick") ? "button" : tag);
    var name = at(el, "aria-label");
    if (!name && el.id && el.ownerDocument) {
      var lab = el.ownerDocument.querySelector('label[for="' + CSS.escape(el.id) + '"]');
      if (lab) name = txt(lab);
    }
    if (!name && tag === "input" && (type === "submit" || type === "button" || type === "reset"))
      name = at(el, "value");
    if (!name && (role === "button" || role === "link")) name = txt(el);
    if (!name) {
      var cell = el.closest ? el.closest("td,th") : null;
      if (cell && cell.previousElementSibling) name = txt(cell.previousElementSibling);
    }
    if (!name) name = at(el, "name") || at(el, "placeholder") || tag;
    return { role: role, name: name, secret: tag === "input" && type === "password" };
  }

  function push(rec) {
    try { top.__cua_human.push(rec); } catch (e) {}
  }

  function attach(doc) {
    if (!doc) return;
    /* Always walk into child frames, even when this document is already hooked:
       a navigation inside a frame replaces that frame's document, and the parent
       document's guard flag must not stop the new one from being hooked. */
    var already = false;
    try { already = !!doc.__cua_rec; doc.__cua_rec = true; } catch (e) { return; }
    if (already) { attachFrames(doc); return; }

    doc.addEventListener("click", function (e) {
      var d = describe(e.target);
      push({
        at: new Date().toISOString(),
        kind: "click",
        role: d.role,
        name: d.name,
        text: txt(e.target),
        frame_url: doc.location ? doc.location.href : "",
        x: e.clientX, y: e.clientY,
      });
    }, true);

    doc.addEventListener("change", function (e) {
      var d = describe(e.target);
      var v = null;
      try { v = d.secret ? "[REDACTED:secret-field]" : String(e.target.value ?? ""); } catch (err) {}
      push({
        at: new Date().toISOString(),
        kind: "input",
        role: d.role,
        name: d.name,
        value: v,
        frame_url: doc.location ? doc.location.href : "",
      });
    }, true);

    doc.addEventListener("submit", function (e) {
      push({
        at: new Date().toISOString(),
        kind: "submit",
        role: "form",
        name: at(e.target, "name") || at(e.target, "action") || "form",
        frame_url: doc.location ? doc.location.href : "",
      });
    }, true);

    attachFrames(doc);
  }

  function attachFrames(doc) {
    var fr = doc.querySelectorAll ? doc.querySelectorAll("iframe,frame") : [];
    for (var i = 0; i < fr.length; i++) {
      try { if (fr[i].contentDocument) attach(fr[i].contentDocument); } catch (e) {}
    }
  }

  attach(document);
  return (top.__cua_human || []).length;
})
